### DISCUSSION TITLE
UDP Server not working

### DISCUSSION THREAD

**Discussion by lump93** [jvde-github/AIS-catcher#29] [Q&A] [answered]
Hi Jasper, thanks a lot for the distribution of this package!
I got good results with my rtl-sdr dongle on my windows laptop and wanted to get the same results on my Raspi.
Installation seems to be fine and I get the same amount of messages into the console, however I dont seem to get the UDP server to work. I tried various IP adresses and ports with my home network and mobile hotspot, but no messages come through.
Then I checked with TShark on the PI and it looks like nothing is beeing send out in the first place.
Is there anything I can do to check if the UDP server is correctly set up?

**Comment by jvde-github** [comment, marked as answer]
Hi, thanks for your question. You could first try to see if it works on your local machine. For example, start AIS-catcher with
```
AIS-catcher  -n -u 127.0.0.1 12345
```
Which should produce AIVDM lines. Try in another terminal on that local machine:
```
nc -l -u 127.0.0.1 12345
```
This listens to the UDP messages coming in. If this does not work, might be something with the internal UDP server of the program.

If you encounter any issues, you can also pipe the UDP lines from screen into netcat and send:
```
AIS-catcher  -n  | nc -u 127.0.0.1 12345
```

Typically what goes wrong is that someone puts the IP address of the Raspi as UDP address, but it should be that of the remote machine. But since you test with tshark might be something different here. 

You can also use a linux command to send a UDP message:
```
echo "\!AIVDM,1,1,,B,45CCLr5000PCAJlMeSWK3l0t0DcJ,0*4D" |nc -u 192.168.1.239 4002
```
This (fake) message is received by my AISdispatcher running @ port 4002 with IP 192.168.1.239.

In this way you can get insight where the issue sits.

What command are you using for AIS-catcher and tshark?

Thanks!

**Reply by lump93** [reply]
Thanks for the quick reply and your suggestions!

The first example works, so the internal UDP server of the program seems to work fine.
My laptop IP adress is 192.168.178.25, so my minimum AIS-catcher command would be the following:
`
AIS-catcher -u 192.168.178.25 10110 -p 40
`

I also tried echoing the AIVDM message via UDP manually like you suggested, but I couldnt catch it with my laptop and OpenCPN (listening to `0.0.0.0:4002`).
tshark probably can be ignored for now, as it also doesnt seem to catch up the existing local UDP communications.

Any other suggestions?
I understand now that this is more likely a network problem, so I can also close this case if you want..

**Comment by jvde-github** [comment]
My command line example was incorrect, for me this works with OpenCPN (I see it popping up in the debug window):
```
echo '!AIVDM,1,1,,B,13aSPvh0000DJ`JMdFTlkHt>0dH4,0*69' |nc -u 192.168.1.239 4002
```

And something basic, like ping the windows PC from your Raspi?
```
ping 192.168.178.25
```

You might try to listen in OpenCPN to  192.168.178.25 instead of 0.0.0.0. Should not make a difference but you never know. Could also try a different (higher) port number to send to/receive on to see if there is some blocking?

**Reply by lump93** [reply]
Yes pinging works, but any kind of UDP communication is not received.
I also tried other applications to rule out a problem with OpenCPN.
I finally tried to receive the same with my phone again (which I thought I'd done before) and this works perfectly.
So it seems I have some problem with my laptops network adapter or so - wondering if it has something to do with IPV6 even though I have obtained an IPV4 address as well.
Anyway in general it seems to work and my goal is to use it with a tablet on a sailing vessel anyway, so I am happy with result for now.
Thanks a lot for your help!

**Comment by moloned** [comment]
I've been struggling with UDP traffic too

In my case the feed to AIScatcher.org works fine using the key supplied https://aiscatcher.org/stations/details/1851

However the UDP feed to AIShub.net does not https://www.aishub.net/stations/3546

I've checked with the systems people in my office and they say they don't filter outbound UDP traffic which tallies with the fact traffic to AIScatcher.org arrives without issue 

I can't for the life of me figure out what the problem is

One issue I came across with the Nooelec SDR I'm using is that aiscatcher doesn't always start up so I do a usbreset to resolve the issue as part of my startup script as follows

SDRreset=$(lsusb | grep DVB-T | awk '{print "sudo ./usbreset /dev/bus/usb/"$2"/"$4}' | sed 's/://') 
echo $SDRreset
$SDRreset

**Reply by jvde-github** [reply]
For aiscatcher.org we are using a TCP connection, not UDP but generally UDP should work fine without issues.
There are some tricks to make it a bit more stable but looking at your AIShub link, no data is flowing at all except for a month ago.

You can check if UDP data is being send by AIS-catcher with (on Linux/RPi):
```
sudo tcpdump udp  
```
In my case, I see lines in the output like:
```
22:44:43.251812 IP raspberrypi.home.42620 > data.aishub.net.XXXX: UDP, length 49
```
This indicates the OS is sending the data. Unfortunately, with UDP, you have no indication that the messages are being received but this is unusual. 

You can try to see if the approach to send to a local address works. Otherwise, I am happy to have a look at the output or config you are using and perhaps the folks at AIShub can advice.  It is unusual.

ps thanks for the tip on the usbreset. Haven't encountered this, but will see if there are is an option to make the program more robust.

**Comment by Phil353556** [comment]
Hi, I'm sharing experience with shipxplorer and UDP connexion.
When an issue arrives, on the server side, not mine, I confirm using the tshark command. The result is
 ICMP 119 Destination unreachable (Port unreachable) "
 
 hope this help.
